src/modules/job-manager/jobtap.c

```diff
@@ -0,0 +1,520 @@
+                              "{s:O s:I s:i s:i s:i s:I s:f}",
+                              "jobspec", job->jobspec_redacted,
+                              "id", job->id,
+                              "userid", job->userid,
+                              "urgency", job->urgency,
+                              "state", job->state,
+                              "priority", job->priority,
+                              "t_submit", job->t_submit) < 0)
+        goto error;
+
+    if (fmt
+        && flux_plugin_arg_vpack (args,
+                                  FLUX_PLUGIN_ARG_IN|FLUX_PLUGIN_ARG_UPDATE,
+                                  fmt, ap) < 0)
+        goto error;
+
+    /*
+     *  Always start with empty OUT args. This allows unpack of OUT
+     *   args to work without error, even if plugin does not set any
+     *   OUT args.
+     */
+    if (flux_plugin_arg_set (args, FLUX_PLUGIN_ARG_OUT, "{}") < 0)
+        goto error;
+
+    return args;
+error:
+    flux_plugin_arg_destroy (args);
+    return NULL;
+}
+
+int jobtap_get_priority (struct jobtap *jobtap,
+                         struct job *job,
+                         int64_t *pprio)
+{
+    int rc = -1;
+    flux_plugin_arg_t *args;
+    va_list ap;
+    int64_t priority = -1;
+
+    if (!jobtap || !job || !pprio) {
+        errno = EINVAL;
+        return -1;
+    }
+    if (!jobtap->plugin) {
+        *pprio = job->urgency;
+        return 0;
+    }
+    /*  N.B.: passing uninitialized ap to args_vcreate, but it should be
+     *   ok, since fmt == NULL, ap will not be used.
+     */
+    if (!(args = jobtap_args_vcreate (jobtap, job, NULL, ap)))
+        return -1;
+    rc = flux_plugin_call (jobtap->plugin, "job.priority.get", args);
+    if (rc == 1) {
+        /*
+         *  A priority.get callback was run. Try to unpack a new priority
+         */
+        if (flux_plugin_arg_unpack (args, FLUX_PLUGIN_ARG_OUT,
+                                    "{s?I}",
+                                    "priority", &priority) < 0) {
+            flux_log (jobtap->ctx->h, LOG_ERR,
+                      "jobtap: %s: job.priority.get: arg_unpack: %s",
+                      jobtap_plugin_name (jobtap->plugin),
+                      flux_plugin_arg_strerror (args));
+            rc = -1;
+        }
+        if (priority == -1) {
+           /*
+            *  A plugin callback was called but didn't provide a
+            *   priority. This could be due to a loaded plugin that is
+            *   not a priority plugin. Therefore take the default action
+            *   and set priority to job->urgency.
+            */
+            priority = job->urgency;
+        }
+        else if (priority == FLUX_JOBTAP_PRIORITY_UNAVAIL) {
+            /*
+             *  Plugin cannot determine priority at this time. Set
+             *   priority to the current job->priority so that a priority
+             *   event is not generated.
+             */
+            priority = job->priority;
```

Comment:
I might be confused. How is it not safe to set priority to the current priority?

If priority is not initialized, then if priority is unavailable, we want to keep it uninitialized so the job stays in the priority state. That is what the current comment is stating.

Reply:
Yeah, I was referring to priority not initialized.  But I now see what you're talking about (and the FLUX_JOB_STATE_SCHED check below it), so I think this is ok.

Although, I am now beginning to wonder if the following is safe:

```
flux job submit --urgency=8 foo
flux job urgency $jobid 10
```

if the priority change in the submit is asynchronous and the priority change in the urgency change is synchronous, could the job have a priority a user wasn't expecting?

Reply:
Hm, good question. (and a good idea for a test).

A priority plugin would have to be written to be able to handle `job.priority.get` callbacks for jobs in the PRIORITY state. If the outstanding priority calculation depends on the current value of the urgency, then in the `job.priority.get` callback, the plugin should update its internal cached version of the urgency so as to not "overwrite" the new urgency value.

O/w, if the `job.priority.get` callback is called just after the plugin sets the initial priority for the job, then it can just update the priority value using the new urgency.

So I think as long as the plugin is written properly, there isn't a problem here, though it is possible I'm forgetting something.